File: scripts/create-tier-snapshots.py
#!/usr/bin/env python3

"""
Create Optimized Tier-Based Snapshots for Neural Weights Hub
Creates snapshots with proper resource allocation and auto-stop disabled
"""

import os
import time
import json
from daytona import (
    Daytona,
    CreateSnapshotParams,
    Image,
    Resources,
    VolumeMount
)

# Initialize Daytona client
daytona = Daytona(
    api_key=os.getenv('DAYTONA_API_KEY'),
    api_url=os.getenv('DAYTONA_SERVER_URL', 'https://app.daytona.io/api'),
    target='us'
)

# Volume IDs from environment
GPT_20B_VOLUME = os.getenv('DAYTONA_GPT_20B_VOLUME_ID', '3d7e7067-1bc1-4094-aaff-9d165fe153e4')
GPT_120B_VOLUME = os.getenv('DAYTONA_GPT_120B_VOLUME_ID', '612103f9-101c-4701-8a33-11f70ab58b1d')

# Snapshot configurations for each tier
TIER_SNAPSHOTS = {
    'neural-weights-free': {
        'name': 'neural-weights-free-v1',
        'description': 'Free tier: 1 CPU, 1GB RAM, GPT-20B access, auto-stop disabled',
        'resources': Resources(cpu=1, memory=1, disk=10),
        'volumes': [
            VolumeMount(volumeId=GPT_20B_VOLUME, mountPath='/models/gpt-20b')
        ],
        'features': ['python', 'jupyter', 'gpt-20b', 'web-terminal']
    },
    'neural-weights-pro': {
        'name': 'neural-weights-pro-v1',
        'description': 'Pro tier: 2 CPU, 4GB RAM, both models, auto-stop disabled',
        'resources': Resources(cpu=2, memory=4, disk=10),
        'volumes': [
            VolumeMount(volumeId=GPT_20B_VOLUME, mountPath='/models/gpt-20b'),
            VolumeMount(volumeId=GPT_120B_VOLUME, mountPath='/models/gpt-120b')
        ],
        'features': ['python', 'jupyter', 'gpt-20b', 'gpt-120b', 'web-terminal', 'enhanced-compute']
    },
    'neural-weights-team': {
        'name': 'neural-weights-team-v1',
        'description': 'Team tier: 4 CPU, 8GB RAM, all models, collaboration tools',
        'resources': Resources(cpu=4, memory=8, disk=10),
        'volumes': [
            VolumeMount(volumeId=GPT_20B_VOLUME, mountPath='/models/gpt-20b'),
            VolumeMount(volumeId=GPT_120B_VOLUME, mountPath='/models/gpt-120b')
        ],
        'features': ['python', 'jupyter', 'gpt-20b', 'gpt-120b', 'web-terminal', 'max-compute', 'collaboration']
    },
    'neural-weights-enterprise': {
        'name': 'neural-weights-enterprise-v1',
        'description': 'Enterprise tier: 4 CPU, 8GB RAM, all features, priority support',
        'resources': Resources(cpu=4, memory=8, disk=10),
        'volumes': [
            VolumeMount(volumeId=GPT_20B_VOLUME, mountPath='/models/gpt-20b'),
            VolumeMount(volumeId=GPT_120B_VOLUME, mountPath='/models/gpt-120b')
        ],
        'features': ['python', 'jupyter', 'gpt-20b', 'gpt-120b', 'web-terminal', 'max-compute', 'collaboration', 'priority-support']
    }
}

def create_tier_snapshot(tier_name, config):
    """Create a snapshot for a specific tier"""
    print(f"\n🚀 Creating {tier_name} snapshot...")
    print(f"   Resources: {config['resources'].cpu} CPU, {config['resources'].memory}GB RAM, {config['resources'].disk}GB disk")
    print(f"   Volumes: {len(config['volumes'])} model volumes")
    print(f"   Features: {', '.join(config['features'])}")
    
    try:
        # Create snapshot with proper resource allocation
        snapshot = daytona.snapshot.create(
            CreateSnapshotParams(
                name=config['name'],
                image=Image.debian_slim("3.12"),  # Python 3.12 base
                resources=config['resources'],
                volumes=config['volumes'],
                labels={
                    'neural-weights/tier': tier_name.replace('neural-weights-', ''),
                    'neural-weights/auto-stop': 'disabled',
                    'neural-weights/features': ','.join(config['features']),
                    'neural-weights/created': time.strftime('%Y-%m-%d %H:%M:%S'),
                    'neural-weights/version': 'v1'
                },
                env_vars={
                    'NEURAL_WEIGHTS_TIER': tier_name.replace('neural-weights-', ''),
                    'PYTHON_VERSION': '3.12',
                    'JUPYTER_ENABLE': 'true',
                    'WEB_TERMINAL_ENABLE': 'true',
                    'AUTO_STOP': 'disabled',
                    'GPT_20B_PATH': '/models/gpt-20b',
                    'GPT_120B_PATH': '/models/gpt-120b' if len(config['volumes']) > 1 else '',
                    'NODE_ENV': 'production'
                }
            ),
            on_logs=lambda log: print(f"   📝 {log}")
        )
        
        print(f"✅ {tier_name} snapshot created: {snapshot.id}")
        return {
            'tier': tier_name,
            'snapshot_id': snapshot.id,
            'name': config['name'],
            'resources': {
                'cpu': config['resources'].cpu,
                'memory': config['resources'].memory,
                'disk': config['resources'].disk
            },
            'volumes': len(config['volumes']),
            'features': config['features'],
            'status': 'success'
        }
        
    except Exception as e:
        print(f"❌ Failed to create {tier_name} snapshot: {str(e)}")
        return {
            'tier': tier_name,
            'name': config['name'],
            'error': str(e),
            'status': 'failed'
        }

def wait_for_snapshot_ready(snapshot_id, max_wait=300):
    """Wait for snapshot to be ready"""
    print(f"⏳ Waiting for snapshot {snapshot_id} to be ready...")
    
    start_time = time.time()
    while time.time() - start_time < max_wait:
        try:
            snapshot = daytona.snapshot.get(snapshot_id)
            if snapshot.state == 'active':
                print(f"✅ Snapshot {snapshot_id} is ready")
                return True
            elif snapshot.state == 'failed':
                print(f"❌ Snapshot {snapshot_id} failed")
                return False
            else:
                print(f"⏳ Snapshot state: {snapshot.state}")
                time.sleep(10)
        except Exception as e:
            print(f"⏳ Checking snapshot status... ({str(e)})")
            time.sleep(10)
    
    print(f"⚠️ Snapshot {snapshot_id} not ready within {max_wait}s")
    return False

def create_all_tier_snapshots():
    """Create all tier-based snapshots"""
    print("🎯 Creating Neural Weights Hub tier-based snapshots")
    print("=" * 60)
    
    results = []
    
    for tier_name, config in TIER_SNAPSHOTS.items():
        result = create_tier_snapshot(tier_name, config)
        results.append(result)
        
        if result['status'] == 'success':
            # Wait for snapshot to be ready before proceeding
            wait_for_snapshot_ready(result['snapshot_id'])
        
        # Brief pause between snapshots
        time.sleep(5)
    
    # Generate summary
    print("\n" + "=" * 60)
    print("📊 SNAPSHOT CREATION SUMMARY")
    print("=" * 60)
    
    successful = [r for r in results if r['status'] == 'success']
    failed = [r for r in results if r['status'] == 'failed']
    
    print(f"✅ Successful: {len(successful)}")
    print(f"❌ Failed: {len(failed)}")
    
    if successful:
        print("\n🎉 Successfully created snapshots:")
        for result in successful:
            print(f"   • {result['name']}: {result['snapshot_id']}")
            print(f"     Resources: {result['resources']['cpu']} CPU, {result['resources']['memory']}GB RAM")
            print(f"     Features: {', '.join(result['features'])}")
    
    if failed:
        print("\n❌ Failed snapshots:")
        for result in failed:
            print(f"   • {result['name']}: {result['error']}")
    
    # Generate environment variables
    print("\n📋 Environment Variables for .env.local:")
    print("# Neural Weights Tier-Based Snapshots")
    for result in successful:
        tier = result['tier'].replace('neural-weights-', '').upper()
        print(f"NEURAL_WEIGHTS_{tier}_SNAPSHOT={result['snapshot_id']}")
    
    # Save results to file
    with open('tier-snapshots-results.json', 'w') as f:
        json.dump(results, f, indent=2)
    
    print(f"\n📄 Results saved to tier-snapshots-results.json")
    return results

if __name__ == "__main__":
    try:
        results = create_all_tier_snapshots()
        
        successful_count = len([r for r in results if r['status'] == 'success'])
        total_count = len(results)
        
        if successful_count == total_count:
            print(f"\n🎉 All {total_count} snapshots created successfully!")
            exit(0)
        else:
            print(f"\n⚠️ {successful_count}/{total_count} snapshots created successfully")
            exit(1)
            
    except Exception as e:
        print(f"\n💥 Script failed: {str(e)}")
        exit(1)

File: scripts/create-tier-snapshots.js
#!/usr/bin/env node

/**
 * Create Optimized Tier-Based Snapshots for Neural Weights Hub
 * Uses Daytona CLI to create snapshots with proper resource allocation and auto-stop disabled
 */

import { execSync } from 'child_process';
import fs from 'fs';

// Tier configurations with appropriate resource allocation
const TIER_CONFIGS = {
  'neural-weights-free': {
    name: 'neural-weights-free-v1',
    description: 'Free tier: 1 CPU, 1GB RAM, GPT-20B access, auto-stop disabled',
    cpu: 1,
    memory: 1,
    disk: 10,
    volumes: ['gpt-20b'],
    autoStop: 0,  // Disabled
    features: ['python', 'jupyter', 'gpt-20b', 'web-terminal']
  },
  'neural-weights-pro': {
    name: 'neural-weights-pro-v1', 
    description: 'Pro tier: 2 CPU, 4GB RAM, both models, auto-stop disabled',
    cpu: 2,
    memory: 4,
    disk: 10,
    volumes: ['gpt-20b', 'gpt-120b'],
    autoStop: 0,  // Disabled
    features: ['python', 'jupyter', 'gpt-20b', 'gpt-120b', 'web-terminal', 'enhanced-compute']
  },
  'neural-weights-team': {
    name: 'neural-weights-team-v1',
    description: 'Team tier: 4 CPU, 8GB RAM, all models, collaboration tools',
    cpu: 4,
    memory: 8,
    disk: 10,
    volumes: ['gpt-20b', 'gpt-120b'],
    autoStop: 0,  // Disabled
    features: ['python', 'jupyter', 'gpt-20b', 'gpt-120b', 'web-terminal', 'max-compute', 'collaboration']
  },
  'neural-weights-enterprise': {
    name: 'neural-weights-enterprise-v1',
    description: 'Enterprise tier: 4 CPU, 8GB RAM, all features, priority support',
    cpu: 4,
    memory: 8,
    disk: 10,
    volumes: ['gpt-20b', 'gpt-120b'],
    autoStop: 0,  // Disabled
    features: ['python', 'jupyter', 'gpt-20b', 'gpt-120b', 'web-terminal', 'max-compute', 'collaboration', 'priority-support']
  }
};

// Volume IDs
const VOLUMES = {
  'gpt-20b': process.env.DAYTONA_GPT_20B_VOLUME_ID || '3d7e7067-1bc1-4094-aaff-9d165fe153e4',
  'gpt-120b': process.env.DAYTONA_GPT_120B_VOLUME_ID || '612103f9-101c-4701-8a33-11f70ab58b1d'
};

function createDockerfile(config) {
  return `FROM python:3.12-slim

# Install system dependencies
RUN apt-get update && apt-get install -y \\
    gcc \\
    gfortran \\
    build-essential \\
    git \\
    curl \\
    vim \\
    nano \\
    htop \\
    && rm -rf /var/lib/apt/lists/*

# Upgrade pip
RUN pip install --upgrade pip

# Install Neural Weights specific packages
RUN pip install \\
    jupyter \\
    jupyterlab \\
    notebook \\
    torch \\
    transformers \\
    accelerate \\
    openai \\
    anthropic \\
    requests \\
    numpy \\
    pandas \\
    matplotlib \\
    seaborn \\
    plotly \\
    scikit-learn \\
    scipy \\
    huggingface_hub \\
    datasets \\
    tokenizers

# Configure Jupyter
RUN jupyter notebook --generate-config
RUN echo "c.NotebookApp.ip = '0.0.0.0'" >> ~/.jupyter/jupyter_notebook_config.py
RUN echo "c.NotebookApp.port = 8888" >> ~/.jupyter/jupyter_notebook_config.py
RUN echo "c.NotebookApp.open_browser = False" >> ~/.jupyter/jupyter_notebook_config.py
RUN echo "c.NotebookApp.allow_root = True" >> ~/.jupyter/jupyter_notebook_config.py

# Create workspace directories
RUN mkdir -p /workspace/notebooks /workspace/scripts /workspace/models

# Create startup script
RUN echo '#!/bin/bash' > /workspace/start-services.sh
RUN echo 'echo "🚀 Starting Neural Weights services..."' >> /workspace/start-services.sh
RUN echo 'nohup jupyter lab --ip=0.0.0.0 --port=8888 --no-browser --allow-root > /workspace/jupyter.log 2>&1 &' >> /workspace/start-services.sh
RUN echo 'echo "✅ Services started!"' >> /workspace/start-services.sh
RUN echo 'echo "📊 Jupyter Lab: http://localhost:8888"' >> /workspace/start-services.sh
RUN chmod +x /workspace/start-services.sh

# Set working directory
WORKDIR /workspace

# Keep container running
CMD ["sleep", "infinity"]`;
}

async function createTierSnapshot(tierName, config) {
  console.log(`\n🚀 Creating ${tierName} snapshot...`);
  console.log(`   Resources: ${config.cpu} CPU, ${config.memory}GB RAM, ${config.disk}GB disk`);
  console.log(`   Volumes: ${config.volumes.join(', ')}`);
  console.log(`   Features: ${config.features.join(', ')}`);

  try {
    // Create Dockerfile
    const dockerfile = createDockerfile(config);
    const dockerfilePath = `./Dockerfile.${tierName}`;
    fs.writeFileSync(dockerfilePath, dockerfile);

    // Build and push snapshot using Daytona CLI
    const command = [
      'daytona snapshot create',
      config.name,
      `--dockerfile ${dockerfilePath}`,
      `--cpu ${config.cpu}`,
      `--memory ${config.memory}`,
      `--disk ${config.disk}`,
      '--auto-stop 0',  // Disable auto-stop
      '--auto-archive 0',  // Disable auto-archive
      '--auto-delete -1'   // Never auto-delete
    ].join(' ');

    console.log(`   📝 Running: ${command}`);
    
    const output = execSync(command, { 
      encoding: 'utf8',
      stdio: 'pipe'
    });

    console.log(`✅ ${tierName} snapshot created successfully`);
    
    // Clean up Dockerfile
    fs.unlinkSync(dockerfilePath);

    // Extract snapshot ID from output (if available)
    const snapshotIdMatch = output.match(/Use '([^']+)' to create/);
    const snapshotId = snapshotIdMatch ? snapshotIdMatch[1] : config.name;

    return {
      tier: tierName,
      snapshotId: snapshotId,
      name: config.name,
      resources: {
        cpu: config.cpu,
        memory: config.memory,
        disk: config.disk
      },
      volumes: config.volumes,
      features: config.features,
      autoStop: config.autoStop,
      status: 'success',
      output: output
    };

  } catch (error) {
    console.error(`❌ Failed to create ${tierName} snapshot:`, error.message);
    
    // Clean up Dockerfile if it exists
    const dockerfilePath = `./Dockerfile.${tierName}`;
    if (fs.existsSync(dockerfilePath)) {
      fs.unlinkSync(dockerfilePath);
    }

    return {
      tier: tierName,
      name: config.name,
      error: error.message,
      status: 'failed'
    };
  }
}

async function createAllTierSnapshots() {
  console.log('🎯 Creating Neural Weights Hub tier-based snapshots');
  console.log('=' * 60);

  const results = [];

  for (const [tierName, config] of Object.entries(TIER_CONFIGS)) {
    const result = await createTierSnapshot(tierName, config);
    results.push(result);

    // Brief pause between snapshots
    await new Promise(resolve => setTimeout(resolve, 5000));
  }

  // Generate summary
  console.log('\n' + '='.repeat(60));
  console.log('📊 SNAPSHOT CREATION SUMMARY');
  console.log('='.repeat(60));

  const successful = results.filter(r => r.status === 'success');
  const failed = results.filter(r => r.status === 'failed');

  console.log(`✅ Successful: ${successful.length}`);
  console.log(`❌ Failed: ${failed.length}`);

  if (successful.length > 0) {
    console.log('\n🎉 Successfully created snapshots:');
    successful.forEach(result => {
      console.log(`   • ${result.name}: ${result.snapshotId}`);
      console.log(`     Resources: ${result.resources.cpu} CPU, ${result.resources.memory}GB RAM`);
      console.log(`     Auto-stop: ${result.autoStop === 0 ? 'DISABLED' : 'ENABLED'}`);
      console.log(`     Features: ${result.features.join(', ')}`);
    });
  }

  if (failed.length > 0) {
    console.log('\n❌ Failed snapshots:');
    failed.forEach(result => {
      console.log(`   • ${result.name}: ${result.error}`);
    });
  }

  // Generate environment variables
  console.log('\n📋 Environment Variables for .env.local:');
  console.log('# Neural Weights Tier-Based Snapshots');
  successful.forEach(result => {
    const tier = result.tier.replace('neural-weights-', '').toUpperCase();
    console.log(`NEURAL_WEIGHTS_${tier}_SNAPSHOT=${result.snapshotId}`);
  });

  // Save results to file
  fs.writeFileSync('tier-snapshots-results.json', JSON.stringify(results, null, 2));
  console.log('\n📄 Results saved to tier-snapshots-results.json');

  return results;
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  createAllTierSnapshots()
    .then(results => {
      const successfulCount = results.filter(r => r.status === 'success').length;
      const totalCount = results.length;

      if (successfulCount === totalCount) {
        console.log(`\n🎉 All ${totalCount} snapshots created successfully!`);
        process.exit(0);
      } else {
        console.log(`\n⚠️ ${successfulCount}/${totalCount} snapshots created successfully`);
        process.exit(1);
      }
    })
    .catch(error => {
      console.error('\n💥 Script failed:', error);
      process.exit(1);
    });
}

export { createAllTierSnapshots, TIER_CONFIGS };
